vyxal/transpile.py

```diff
@@ -0,0 +1,164 @@
+from vyxal.structure import *
+
+
+def lambda_wrap(branch: List[Structure]) -> Lambda:
+    """
+    Turns a List of structures into a single lambda  Useful
+    for dealing with the functions of modifiers. Note that single
+    elements pass their arity on to the lambda
+
+    Para
+    """
+
+    if len(branch) == 1:
+        if isinstance(branch[0], GenericStatement):
+            return Lambda([branch[0]])
+            # TODO: Actually get arity
+            # of the element and make that the arity of the lambda being
+            # returned. This'll be possible once we actually get the
+            # command dictionary established
+        elif isinstance(branch[0], Lambda):
+            return branch[0]
+        else:
+            return Lambda(branch)
+    else:
+        return Lambda(branch)
+
+
+def transpile(program: str) -> str:
+    return transpile_ast(parse.parse(lexer.tokenise(program)))
+
+
+def transpile_ast(program: List[Structure], indent=0) -> str:
+    """
+    Transpile a given program (as a parsed list of structures) into Python
+    """
+    if not program:
+        return helpers.indent_str("pass", indent)
+    return "\n".join(transpile_single(struct, indent=indent) for struct in program)
+
+
+def transpile_single(token_or_struct: Union[Token, Structure], indent: int) -> str:
+    if isinstance(token_or_struct, Token):
+        return transpile_token(token_or_struct, indent)
+    elif isinstance(token_or_struct, Structure):
+        return transpile_structure(token_or_struct, indent)
+    raise ValueError(
+        f"Input must be a Token or Structure, was {type(token_or_struct).__name__}: {token_or_struct}"
+    )
+
+
+def transpile_token(token: Token, indent: int) -> str:
+    from vyxal.helpers import indent_str
+
+    if token.name == TokenType.STRING:
+        return indent_str(f"stack.append('{token.value}')", indent)
+    elif token.name == TokenType.NUMBER:
+        return indent_str(f"stack.append({token.value})", indent)
+    elif token.name == TokenType.NAME:
+        pass
+    elif token.name == TokenType.GENERAL:
+        pass  # return elements.elements.get(token.value) @lyxal is this right?
+    elif token.name == TokenType.COMPRESSED_NUMBER:
+        pass
+    elif token.name == TokenType.COMPRESSED_STRING:
+        pass
+    elif token.name == TokenType.VARIABLE_GET:
+        pass
+    elif token.name == TokenType.VARIABLE_SET:
+        pass
+    raise ValueError(f"Bad token: {token}")
+
+
+def transpile_structure(struct: Structure, indent: int) -> str:
+    """
+    Transpile a single structure.
+    """
+    from vyxal.helpers import indent_str
+
+    if isinstance(struct, GenericStatement):
+        return transpile_single(struct.branches[0], indent)
+    if isinstance(struct, IfStatement):
+        # Add an empty "branch" as the first condition
+        branches = [""] + struct.branches
+        # Holds indentation level as elifs will be nested inside the else part
```

Comment:
is there a reason not to use actual Python `elif`?

Reply:
Because we need to run the code for the condition, pop it, and push it into `context_values` each time. I'll look into making it simpler, though.

Reply:
Ah. Maybe `elif do_setup() and condition():` would work?

Reply:
Mmm, I don't know how that'd work with conditions that are compiled to multiline code (if nested inside if), but I'll look into it.